[quant] Input-Weight Equalization - Conv observer support

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61350
* #61287
* #61286
* **#61285**

Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

angelayi committed:
[quant] Input-Weight Equalization - Conv observer support

Summary: Modifies observers to support conv layers.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557041).

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -61,9 +61,8 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
-        if not (x_orig.ndim == 2):
-            raise ValueError("InputEqualizationObserver only supports Linear layers")
+        if not (x_orig.ndim == 2 or x_orig.ndim == 4):
+            raise ValueError("InputEqualizationObserver only supports Linear and Conv layers")
```
We probably need to be able to support conv with 1/2/3 dim as well. Currently looks like this only supports conv2d.

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -55,44 +55,64 @@
+
+        return (mins, maxs)
+
+    @given(ndim=st.sampled_from((2, 4)),
```
maybe add tests for 2 to 5 (linear, conv1d, 2d, 3d)

jerryzh168 reviewed (approved):
please add the tests for conv1d and 3d as well

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557041).

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - Conv observer support"


Summary: Modifies observers to support conv layers and tests to make sure that the observers are returning the expected values for conv inputs.

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_eq_observer`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29557041](https://our.internmc.facebook.com/intern/diff/D29557041)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29557041).
